File: Templates/Range Queries/BIT.java

class BIT {

	// one indexed as operations on index 0 does not work
	int[] BIT;
	int numE;

	BIT(int[] initalData) {
		BIT = new int[initalData.length];
		numE = initalData.length - 1;
		for (int i = 1; i < initalData.length; i++) {
			update(i, initalData[i]);
		}
	}

	void update(int ind, int delta) {
		while (ind <= numE) {
			BIT[ind] += delta;
			ind += (ind & -ind);
		}
	}

	int query(int ind) {
		int sum = 0;
		while (ind > 0) {
			sum += BIT[ind];
			ind -= (ind & -ind);
		}
		return sum;
	}
}

File: C++ Templates/Ranges/BIT.cpp
struct BIT {
    int tr[MAXN];

    void u(int i, int d) {
        while (i < MAXN) {
            tr[i] += d;
            i += (i & - i);
        }
    }

    int q(int i) {
        int s = 0;
        while (i > 0) {
            s += tr[i];
            i -= (i & -i);
        }
        return s;
    }
};
